src/NumericalAlgorithms/Integration/GslQuadAdaptive.hpp

```diff
@@ -0,0 +1,295 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <gsl/gsl_integration.h>
+#include <memory>
+
+#include "Utilities/Gsl.hpp"
+
+/// \cond
+namespace PUP {
+class er;
+}  // namespace PUP
+/// \endcond
+
+/// Numerical integration algorithms
+namespace integration {
+
+namespace detail {
+
+// The GSL functions require the integrand to have this particular function
+// signature. In particular, any extra parameters to the functions must be
+// passed as a void* and re-interpreted appropriately.
+template <typename IntegrandType>
+double integrand(const double x, void* const params) {
+  const auto function = reinterpret_cast<IntegrandType*>(params);
+  return (*function)(x);
+}
+
+class GslQuadAdaptiveImpl {
+ public:
+  explicit GslQuadAdaptiveImpl(size_t max_intervals) noexcept;
+
+  GslQuadAdaptiveImpl() = default;
+  GslQuadAdaptiveImpl(const GslQuadAdaptiveImpl&) = delete;
+  GslQuadAdaptiveImpl& operator=(const GslQuadAdaptiveImpl&) = delete;
+  GslQuadAdaptiveImpl(GslQuadAdaptiveImpl&&) noexcept = default;
+  GslQuadAdaptiveImpl& operator=(GslQuadAdaptiveImpl&& rhs) noexcept = default;
+  ~GslQuadAdaptiveImpl() noexcept = default;
+
+  void pup(PUP::er& p) noexcept;  // NOLINT(google-runtime-references)
+
+ private:
+  struct gsl_integration_workspace_deleter {
+    void operator()(gsl_integration_workspace* workspace) const noexcept;
+  };
+
+ protected:
+  template <typename IntegrandType>
+  gsl_function* gsl_integrand(IntegrandType&& integrand) const noexcept {
+    gsl_integrand_.function = &detail::integrand<IntegrandType>;
+    gsl_integrand_.params = &integrand;
+    return &gsl_integrand_;
+  }
+
+  size_t max_intervals_;
+  std::unique_ptr<gsl_integration_workspace, gsl_integration_workspace_deleter>
+      workspace_;
+
+ private:
+  void initialize() noexcept;
+
+  mutable gsl_function gsl_integrand_;
+};
+
+void check_status_code(int status_code);
+
+void disable_gsl_error_handling() noexcept;
+
+}  // namespace detail
+
+/// Each type specifies which algortihm from GSL should be used. It should be
+/// chosen according to the problem.
+enum class GslIntegralType {
+  StandardGaussKronrod  /// gsl_integration_qag()
+  ,
+  IntegrableSingularitiesPresent  /// gsl_integration_qags()
+  ,
+  IntegrableSingularitiesKnown  /// gsl_integration_qagp()
+  ,
+  InfiniteInterval  /// gsl_integration_qagi()
+  ,
+  UpperBoundaryInfinite  /// gsl_integration_qagiu()
+  ,
+  LowerBoundaryInfinite  /// gsl_integration_qagil()
+};
+
+/*!
+ * \ingroup NumericalAlgorithmsGroup
```

Comment:
Let's not stick this into the `NumericalAlgorithmsGroup` and instead add the namespace. I think otherwise the `NumericalAlgorithmsGroup` will get more crowded than it already is. In hindsight we probably started doing only the namespaces from the beginning.